File: WinForms/VB/EditorsMainDemo/Modules/ModuleTokenEdit.vb
Imports DevExpress.XtraEditors.Demos.Data
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid
Imports System
Imports System.Data
Imports System.Linq
Imports System.Windows.Forms

Namespace DevExpress.XtraEditors.Demos

    Public Partial Class ModuleTokenEdit
        Inherits TutorialControl

        Private CountriesTokenEdit As RepositoryItemTokenEdit

        Private ProducersTokenEdit As RepositoryItemTokenEdit

        Private GenresTokenEdit As RepositoryItemTokenEdit

        Const InitialEditorRowHandle As Integer = 17

        Shared Sub New()
            Call MatchFunction.Register()
        End Sub

        Public Sub New()
            Call MoviesData.Load()
            InitializeComponent()
            InitControls()
        End Sub

        Private Sub InitControls()
            gridControl.DataSource = Movies
            InitTokenEditGenres()
            InitTokenEditProducers()
            InitTokenEditCountries()
            InitRepositoryItemTokens()
            filterPanel.SetFilterCriteriaBinding(gridView)
            gridView.FocusedRowHandle = InitialEditorRowHandle
        End Sub

        Private lockFilterSync As Integer = 0

        Private Sub GridView_ColumnFilterChanged(ByVal sender As Object, ByVal e As EventArgs)
            lockFilterSync += 1
            If colGenres.FilterInfo.Equals(ColumnFilterInfo.Empty) Then tokenEditGenres.EditValue = MovieGenre.None
            If colCountries.FilterInfo.Equals(ColumnFilterInfo.Empty) Then tokenEditCountries.EditValue = Nothing
            If colProducers.FilterInfo.Equals(ColumnFilterInfo.Empty) Then tokenEditProducers.EditValue = Nothing
            lockFilterSync -= 1
        End Sub

        ' <tokenEditGenres>
        Private Sub InitTokenEditGenres()
            tokenEditGenres.Properties.Tokens.AddEnum(GetType(MovieGenre))
        End Sub

        ' </tokenEditGenres>
        ' <tokenEditProducers>
        Private Sub InitTokenEditProducers()
            tokenEditProducers.Properties.BeginUpdate()
            tokenEditProducers.Properties.DataSource = Producers
            tokenEditProducers.Properties.DisplayMember = "Name"
            tokenEditProducers.Properties.ValueMember = "Id"
            tokenEditProducers.Properties.EndUpdate()
        End Sub

        ' </tokenEditProducers>
        ' <tokenEditCountries>
        Private Sub InitTokenEditCountries()
            tokenEditCountries.Properties.Tokens.BeginUpdate()
            tokenEditCountries.Properties.Tokens.AddToken("France")
            tokenEditCountries.Properties.Tokens.AddToken("Germany")
            tokenEditCountries.Properties.Tokens.AddToken("Italy")
            tokenEditCountries.Properties.Tokens.AddToken("Japan")
            tokenEditCountries.Properties.Tokens.AddToken("Spain")
            tokenEditCountries.Properties.Tokens.AddToken("UK")
            tokenEditCountries.Properties.Tokens.AddToken("USA")
            tokenEditCountries.Properties.Tokens.EndUpdate()
        End Sub

        ' </tokenEditCountries>
        Private Sub InitRepositoryItemTokens()
            CountriesTokenEdit = New RepositoryItemTokenEdit()
            ProducersTokenEdit = New RepositoryItemTokenEdit()
            GenresTokenEdit = New RepositoryItemTokenEdit()
            CountriesTokenEdit.Tokens.AddRange(tokenEditCountries.Properties.Tokens)
            ProducersTokenEdit.DataSource = tokenEditProducers.Properties.DataSource
            ProducersTokenEdit.DisplayMember = tokenEditProducers.Properties.DisplayMember
            ProducersTokenEdit.ValueMember = tokenEditProducers.Properties.DisplayMember
            GenresTokenEdit.Tokens.AddRange(tokenEditGenres.Properties.Tokens)
            GenresTokenEdit.Tokens.AddToken("None")
        End Sub

        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            lockFilterSync += 1
            tokenEditGenres.EditValue = MovieGenre.Action Or MovieGenre.Music Or MovieGenre.Musical Or MovieGenre.Comedy
            SetTaggedBy(tokenEditGenres.SelectedItems, colGenres)
            tokenEditCountries.EditValue = "USA, Italy, Spain, France"
            SetTaggedBy(tokenEditCountries.SelectedItems, colCountries)
            lockFilterSync -= 1
        End Sub

        Protected Overrides ReadOnly Property WhatsThisCodeFileNames As String()
            Get
                Return New String() {"ModuleTokenEdit"}
            End Get
        End Property

        Protected Overrides ReadOnly Property WhatsThisXMLFileName As String
            Get
                Return "tokenEdit"
            End Get
        End Property

        Private Sub tokenEditCountries_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            If lockFilterSync = 0 Then SetTaggedBy(tokenEditCountries.SelectedItems, colCountries)
        End Sub

        Private Sub tokenEditGenres_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            If lockFilterSync = 0 Then SetTaggedBy(tokenEditGenres.SelectedItems, colGenres)
        End Sub

        Private Sub tokenEditProducers_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            If lockFilterSync = 0 Then SetTaggedBy(tokenEditProducers.SelectedItems, colProducers)
        End Sub

        Private Sub SetTaggedBy(ByVal tokens As TokenEditSelectedItemCollection, ByVal column As GridColumn)
            If Not tokens.Any() Then
                column.ClearFilter()
            Else
                column.FilterInfo = New ColumnFilterInfo(MatchFunction.Criteria(column.FieldName, tokens.[Select](Function(x) x.Description)))
            End If
        End Sub

        Protected Overrides Sub DoShow()
            If Parent IsNot Nothing Then Parent.Padding = Padding.Empty
            MyBase.DoShow()
        End Sub

        Private Sub gridView_CustomRowCellEditForEditing(ByVal sender As Object, ByVal e As CustomRowCellEditEventArgs)
            If e.Column Is colProducers Then
                e.RepositoryItem = ProducersTokenEdit
            ElseIf e.Column Is colCountries Then
                e.RepositoryItem = CountriesTokenEdit
            ElseIf e.Column Is colGenres Then
                e.RepositoryItem = GenresTokenEdit
            End If
        End Sub

        Private Sub ClearFilter_Click(ByVal sender As Object, ByVal e As EventArgs)
            tokenEditCountries.EditValue = Nothing
            tokenEditGenres.EditValue = Nothing
            tokenEditProducers.EditValue = Nothing
        End Sub

        Protected Overrides Sub DoVisibleChanged(ByVal visible As Boolean)
            MyBase.DoVisibleChanged(visible)
            If visible Then
                gridControl.ForceInitialize()
                gridControl.BeginInvoke(New Action(Sub() gridView.ShowEditForm()))
            End If
        End Sub

        Private Sub gridView_EditFormPrepared(ByVal sender As Object, ByVal e As EditFormPreparedEventArgs)
            For Each edit In e.BindableControls.Cast(Of BaseEdit)()
                edit.Properties.AutoHeight = True
            Next
        End Sub
    End Class
End Namespace

File: WinForms/CS/EditorsMainDemo/Modules/ModuleTokenEdit.cs
using DevExpress.XtraEditors.Demos.Data;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;
using System;
using System.Data;
using System.Linq;
using System.Windows.Forms;

namespace DevExpress.XtraEditors.Demos {
    public partial class ModuleTokenEdit : TutorialControl {
        RepositoryItemTokenEdit CountriesTokenEdit;
        RepositoryItemTokenEdit ProducersTokenEdit;
        RepositoryItemTokenEdit GenresTokenEdit;
        const int InitialEditorRowHandle = 17;

        static ModuleTokenEdit() {
            MatchFunction.Register();
        }
        public ModuleTokenEdit() {
            MoviesData.Load();
            InitializeComponent();
            InitControls();
        }
        void InitControls() {
            gridControl.DataSource = MoviesData.Movies;
            InitTokenEditGenres();
            InitTokenEditProducers();
            InitTokenEditCountries();
            InitRepositoryItemTokens();
            filterPanel.SetFilterCriteriaBinding(gridView);
            gridView.FocusedRowHandle = InitialEditorRowHandle;
        }
        int lockFilterSync = 0;
        void GridView_ColumnFilterChanged(object sender, EventArgs e) {
            lockFilterSync++;
            if(colGenres.FilterInfo.Equals(ColumnFilterInfo.Empty))
                tokenEditGenres.EditValue = MoviesData.MovieGenre.None;
            if(colCountries.FilterInfo.Equals(ColumnFilterInfo.Empty))
                tokenEditCountries.EditValue = null;
            if(colProducers.FilterInfo.Equals(ColumnFilterInfo.Empty))
                tokenEditProducers.EditValue = null;
            lockFilterSync--;
        }
        // <tokenEditGenres>
        void InitTokenEditGenres() {
            tokenEditGenres.Properties.Tokens.AddEnum(typeof(MoviesData.MovieGenre));
        }
        // </tokenEditGenres>
        // <tokenEditProducers>
        void InitTokenEditProducers() {
            tokenEditProducers.Properties.BeginUpdate();
            tokenEditProducers.Properties.DataSource = MoviesData.Producers;
            tokenEditProducers.Properties.DisplayMember = "Name";
            tokenEditProducers.Properties.ValueMember = "Id";
            tokenEditProducers.Properties.EndUpdate();
        }
        // </tokenEditProducers>
        // <tokenEditCountries>
        void InitTokenEditCountries() {
            tokenEditCountries.Properties.Tokens.BeginUpdate();
            tokenEditCountries.Properties.Tokens.AddToken("France");
            tokenEditCountries.Properties.Tokens.AddToken("Germany");
            tokenEditCountries.Properties.Tokens.AddToken("Italy");
            tokenEditCountries.Properties.Tokens.AddToken("Japan");
            tokenEditCountries.Properties.Tokens.AddToken("Spain");
            tokenEditCountries.Properties.Tokens.AddToken("UK");
            tokenEditCountries.Properties.Tokens.AddToken("USA");
            tokenEditCountries.Properties.Tokens.EndUpdate();
        }
        // </tokenEditCountries>
        void InitRepositoryItemTokens() {
            CountriesTokenEdit = new RepositoryItemTokenEdit();
            ProducersTokenEdit = new RepositoryItemTokenEdit();
            GenresTokenEdit = new RepositoryItemTokenEdit();

            CountriesTokenEdit.Tokens.AddRange(tokenEditCountries.Properties.Tokens);

            ProducersTokenEdit.DataSource = tokenEditProducers.Properties.DataSource;
            ProducersTokenEdit.DisplayMember = tokenEditProducers.Properties.DisplayMember;
            ProducersTokenEdit.ValueMember = tokenEditProducers.Properties.DisplayMember;

            GenresTokenEdit.Tokens.AddRange(tokenEditGenres.Properties.Tokens);
            GenresTokenEdit.Tokens.AddToken("None");
        }
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            lockFilterSync++;
            tokenEditGenres.EditValue = MoviesData.MovieGenre.Action | MoviesData.MovieGenre.Music | MoviesData.MovieGenre.Musical | MoviesData.MovieGenre.Comedy;
            SetTaggedBy(tokenEditGenres.SelectedItems, colGenres);
            tokenEditCountries.EditValue = "USA, Italy, Spain, France";
            SetTaggedBy(tokenEditCountries.SelectedItems, colCountries);
            lockFilterSync--;
        }
        protected override string[] WhatsThisCodeFileNames {
            get { return new string[] { "ModuleTokenEdit" }; }
        }
        protected override string WhatsThisXMLFileName {
            get { return "tokenEdit"; }
        }
        void tokenEditCountries_EditValueChanged(object sender, EventArgs e) {
            if(lockFilterSync == 0)
                SetTaggedBy(tokenEditCountries.SelectedItems, colCountries);
        }
        void tokenEditGenres_EditValueChanged(object sender, EventArgs e) {
            if(lockFilterSync == 0)
                SetTaggedBy(tokenEditGenres.SelectedItems, colGenres);
        }
        void tokenEditProducers_EditValueChanged(object sender, EventArgs e) {
            if(lockFilterSync == 0)
                SetTaggedBy(tokenEditProducers.SelectedItems, colProducers);
        }
        void SetTaggedBy(TokenEditSelectedItemCollection tokens, GridColumn column) {
            if(!tokens.Any())
                column.ClearFilter();
            else
                column.FilterInfo = new ColumnFilterInfo(MatchFunction.Criteria(column.FieldName, tokens.Select(x => x.Description)));
        }
        protected override void DoShow() {
            if(Parent != null)
                Parent.Padding = Padding.Empty;
            base.DoShow();
        }
        void gridView_CustomRowCellEditForEditing(object sender, CustomRowCellEditEventArgs e) {
            if(e.Column == colProducers)
                e.RepositoryItem = ProducersTokenEdit;
            else if(e.Column == colCountries)
                e.RepositoryItem = CountriesTokenEdit;
            else if(e.Column == colGenres)
                e.RepositoryItem = GenresTokenEdit;
        }
        void ClearFilter_Click(object sender, EventArgs e) {
            tokenEditCountries.EditValue = null;
            tokenEditGenres.EditValue = null;
            tokenEditProducers.EditValue = null;
        }
        protected override void DoVisibleChanged(bool visible) {
            base.DoVisibleChanged(visible);
            if(visible) {
                gridControl.ForceInitialize();
                gridControl.BeginInvoke(new Action(() => {
                    gridView.ShowEditForm();
                }));
            }
        }

        private void gridView_EditFormPrepared(object sender, EditFormPreparedEventArgs e) {
            foreach(var edit in e.BindableControls.Cast<BaseEdit>())
                edit.Properties.AutoHeight = true;
        }
    }
}
